packages/plugins/src/foundry/utils/index.ts

```diff
@@ -89,40 +123,116 @@ export const makeGetConfigArtifacts = (
         time: fs.statSync(`${buildInfoPath}/${fileName}`).mtime.getTime(),
       }))
       .sort((a, b) => b.time - a.time)
-      .map(async (file) => {
-        return JSON.parse(
-          await readFileAsync(join(buildInfoFolder, file.name), 'utf8')
-        )
-      })
 
-    const buildInfos = await Promise.all(buildInfoPromises)
+    // Read all of the new/modified files and update the cache to reflect the changes
+    // Keep an in memory cache of the read files so we don't have to read them again later
+    const localBuildInfoCache = {}
+    await Promise.all(
+      buildInfoFileNames
+        .filter((file) => buildInfoCache[file.name]?.time !== file.time)
+        .map(async (file) => {
+          const buildInfo = JSON.parse(
+            await readFileAsync(join(buildInfoFolder, file.name), 'utf8')
+          )
+
+          // We keep track of the last modified time in each build info file so we can check the file if it has changed.
+          // We also keep track of all the contracts output by each build info file, so we can easily look up the required file for each source name
+          buildInfoCache[file.name] = {
+            name: file.name,
+            time: file.time,
+            contracts: Object.keys(buildInfo.output.contracts),
+          }
+
+          localBuildInfoCache[file.name] = buildInfo
+        })
+    )
 
-    const configArtifactPromises = Object.entries(contractConfigs).map(
-      async ([referenceName, contractConfig]) => {
-        const artifact = await getContractArtifact(
-          contractConfig.contract,
-          artifactFolder
-        )
-        const buildInfo = getBuildInfo(buildInfos, artifact.sourceName)
+    // Look through the cache, read all the contract artifacts, and find all of the build info files names required for the passed in contract config
+    const toReadFiles: string[] = []
+    const resolved = await Promise.all(
+      Object.entries(contractConfigs).map(
+        async ([referenceName, contractConfig]) => {
+          const artifact = await getContractArtifact(
+            contractConfig.contract,
+            artifactFolder
+          )
+
+          // Just make sure the files are sorted by time
+          const sortedCachedFiles = Object.values(buildInfoCache).sort(
+            (a, b) => b.time - a.time
+          )
+          // Look through the cahce for the first build info file that contains the contract
+          for (const file of sortedCachedFiles) {
+            if (file.contracts.includes(artifact.sourceName)) {
```

Comment:
Should we keep track of fully qualified names (source name + contract name) instead of just the source name in `file.contracts`? Then we'd be able to compare this against the artifact's fully qualified name here. It feels a little safer since I'm not confident about how the solidity compiler behaves in edge cases involving multiple contracts in the same source file

Reply:
I don't think this should be necessary here because we're fetching the source name from the contract artifact, so I don't see a strong reason why we would need the fully qualified name here. As long as we have the correct contract artifact, we'll have the correct source name here as well.

This seems related to the other bug you pointed out with how we're fetching contract artifacts where there's some ambiguity if the user has multiple contracts of the same name. That ambiguity does affect this code since we're assuming that the contract artifact is correct, but I don't think we should try to address the ambiguity here. I think we should just resolve the root issue.